File: notes/0239/0239.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        if not nums: return []
        dp = [-float('inf')]*(len(nums)-k+1)
        dp[0] = max(nums[:k])
        for i in range(1, len(dp)):
            if nums[i+k-1] > dp[i-1]:
                dp[i] = nums[i+k-1]
            if nums[i+k-1] < dp[i-1] and nums[i-1] < dp[i-1]:
                dp[i] = dp[i-1]
            else:
                dp[i] = max(nums[i:i+k])
        return dp

File: notes/0239/0239.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        deque<int> deq;
        vector<int> ans;
        int size = nums.size();
        
        for (int i=0; i<size; i++) {
            while(!deq.empty() && nums[deq.back()] <= nums[i]) deq.pop_back();
            deq.push_back(i);
            if (i-k+1 >= 0) ans.push_back(nums[deq.front()]);
            if (i-k+1 >= deq.front()) deq.pop_front();
        }
        return ans;
    }
};
